server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -2272,6 +2308,80 @@ private void writeAtomic(BlobContainer container, final String blobName, final B
         }
     }
 
+    @Override
+    public void snapshotShardWithRemoteStore(
+        Store store,
+        MapperService mapperService,
+        SnapshotId snapshotId,
+        IndexId indexId,
+        IndexCommit snapshotIndexCommit,
+        String shardStateIdentifier,
+        IndexShardSnapshotStatus snapshotStatus,
+        Version repositoryMetaVersion,
+        Map<String, Object> userMetadata,
+        String primaryTermGenerationPair,
+        ActionListener<String> listener
+    ) {
+        if (isReadOnly()) {
+            listener.onFailure(new RepositoryException(metadata.name(), "cannot snapshot shard on a readonly repository"));
+            return;
+        }
+        final ShardId shardId = store.shardId();
+        final long startTime = threadPool.absoluteTimeInMillis();
+        try {
+            final String generation = snapshotStatus.generation();
+            logger.info("[{}] [{}] snapshot to [{}] [{}] ...", shardId, snapshotId, metadata.name(), generation);
+            final BlobContainer shardContainer = shardContainer(indexId, shardId);
+
+            long indexTotalFileSize = 0;
+            Collection<String> fileNames = snapshotIndexCommit.getFileNames();
+            for (String fileName: fileNames) {
+                indexTotalFileSize += store.getMetadata(snapshotIndexCommit).get(fileName).length();
+            }
+            int indexTotalNumberOfFiles = fileNames.size();
+
+            snapshotStatus.moveToStarted(
+                startTime,
+                0, // incremental File Count is zero as we are storing the data as part of remote store.
+                indexTotalNumberOfFiles,
+                0, // incremental File Size is zero as we are storing the data as part of remote store.
+                indexTotalFileSize
+            );
+
+            final String indexGeneration;
+            indexGeneration = UUIDs.randomBase64UUID() + ShardGenerations.REMOTE_STORE_INTEROP_SHARD_GEN_SUFFIX;
+            final IndexShardSnapshotStatus.Copy lastSnapshotStatus = snapshotStatus.moveToFinalize(snapshotIndexCommit.getGeneration());
+
+            // now create and write the commit point
+            logger.trace("[{}] [{}] writing shard snapshot file", shardId, snapshotId);
+            try {
+                REM_STORE_BASED_INDEX_SHARD_SNAPSHOT_FORMAT.write(
```

Comment:
What will happen for cases when snapshot is triggered multiple times but the data in the Remote store backed index has not changed and that means the metadata has also not changed ? Do we need to write this info again for this new snapshot ? Also am wondering there is `BlobStoreIndexShardSnapshots` as well for regular index snapshot, do we need that for remote store index for any other snapshot API or operation ?

Reply:
with each snapshot, we anyway create a new commit, so remote store backed index shard will always have some files which need to be moved to remote store and therefore we will have new metadata every time.
BlobStoreIndexShardSnapshots was helpful because we were keeping incremental diffs in snapshots. haven't seen its requirement even for other operations.